![Imgur](https://i.imgur.com/5pXzCIu.png)

# Data Science va Sun'iy Intellekt Praktikum

## 3-MODUL. MA’LUMOTLARGA ISHLOV BERISH

### 3.2 Ma'lumotlarni tozalash va tayyorlash

## Boshqa foydali funksiyalar

### No-odatiy qiymatlarni topish

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("https://github.com/anvarnarz/praktikum_datasets/raw/main/MOCK_USER_DATA.csv", index_col=0)
df.head()

first_name   last_name                           email       gender   age
id                                                                          
1     Dorothy     Verryan               dverryan0@who.int   Non-binary     1
2     Pietrek     Scanlin          pscanlin1@yolasite.com         Male    91
3      Lilith    Sarjeant  lsarjeant2@creativecommons.org  Genderfluid  1233
4        Knox  Cunningham    kcunningham3@yellowpages.com  Genderfluid     6
5    Mohammed  Habbershon    mhabbershon4@sourceforge.net       Female    48

Yuqoridagi jadvlada odamlarning yoshi jamlangan. Keling jadval haqida qisqa ma'lumot olamiz:

In [3]:
df.describe()

age
count   100.000000
mean    145.660000
std     565.631411
min     -85.000000
25%      22.750000
50%      53.500000
75%      84.250000
max    4444.000000

Nimadur xato: `max` yosh `4444` deb ko'rsatayapti, tabiiyku bunday bo'lishi mumkin emas. `min` da ham manfiy qiymat ko'rishimiz mumkin, bu ham xato. Demak jadvalimizda noto'g'ri qiymatlar ham kirib qolgan.

Keling, boshlanishiga katta qiymatlarni filtrlab olamiz, yaqin tarixda odam yoshi maksimum `123` dan oshmagnini bilgan holda, keling shu qiymatdan katta qatorlarni topamiz:

In [4]:
df[df.age>123]

first_name  last_name                           email       gender   age
id                                                                          
3       Lilith   Sarjeant  lsarjeant2@creativecommons.org  Genderfluid  1233
39      Godard      Miere               gmiere12@1688.com   Polygender  3434
47       Cleve  Sparshatt      csparshatt1a@economist.com  Genderfluid   333
49       Trace   Stockney          tstockney1c@flavors.me       Female   344
65      Eloise    Snoddon    esnoddon1s@elegantthemes.com  Genderfluid   332
80       Gabie     Rothon              grothon27@ehow.com  Genderqueer  4444
100      Mable    Toffano          mtoffano2r@myspace.com     Bigender   233

Qatorlarning indeksini ham ajratib olishimiz mumkin:

In [5]:
idx = df[df.age>123].index
idx

Index([3, 39, 47, 49, 65, 80, 100], dtype='int64', name='id')

Manfiy qiymatlarni ham yuqoridagi usul bilan filtrlab olishimiz mumkin:

In [6]:
df[df.age<0]

first_name  last_name                           email       gender  age
id                                                                        
11     Elijah      Heine                eheinea@imdb.com     Bigender  -85
19      Luisa  Bradborne            lbradbornei@fema.gov       Female  -75
21      Levin     Worner                lwornerk@ihg.com  Genderqueer  -43
33   Ebenezer    Greeves           egreevesw@myspace.com         Male  -26
38       Hana   Langwade  hlangwade11@washingtonpost.com  Genderqueer  -42
90  Zachariah      Meins            zmeins2h@weather.com         Male  -69

Ikki filtrni jamlab yozish ham mumkin:

In [7]:
df[(df.age>123) | (df.age<0)]

first_name  last_name                           email       gender   age
id                                                                          
3       Lilith   Sarjeant  lsarjeant2@creativecommons.org  Genderfluid  1233
11      Elijah      Heine                eheinea@imdb.com     Bigender   -85
19       Luisa  Bradborne            lbradbornei@fema.gov       Female   -75
21       Levin     Worner                lwornerk@ihg.com  Genderqueer   -43
33    Ebenezer    Greeves           egreevesw@myspace.com         Male   -26
38        Hana   Langwade  hlangwade11@washingtonpost.com  Genderqueer   -42
39      Godard      Miere               gmiere12@1688.com   Polygender  3434
47       Cleve  Sparshatt      csparshatt1a@economist.com  Genderfluid   333
49       Trace   Stockney          tstockney1c@flavors.me       Female   344
65      Eloise    Snoddon    esnoddon1s@elegantthemes.com  Genderfluid   332
80       Gabie     Rothon              grothon27@ehow.com  Genderqueer  4444
90   Zachariah      Meins            zmeins2h@weather.com         Male   -69
100      Mable    Toffano          mtoffano2r@myspace.com     Bigender   233

Qatorlar bilan qanday yo'l tutish sizga bog'liq. Manfiy qiymatlarni musbatga o'tkazishimiz mumkin, buning uchun `numpy` dagi `abs` funksiyasini qo'llaymiz:

In [8]:
df['age'] = df['age'].apply(np.abs)
df.head(15)

first_name   last_name                           email       gender   age
id                                                                          
1     Dorothy     Verryan               dverryan0@who.int   Non-binary     1
2     Pietrek     Scanlin          pscanlin1@yolasite.com         Male    91
3      Lilith    Sarjeant  lsarjeant2@creativecommons.org  Genderfluid  1233
4        Knox  Cunningham    kcunningham3@yellowpages.com  Genderfluid     6
5    Mohammed  Habbershon    mhabbershon4@sourceforge.net       Female    48
6       Glynn      Ebbing          gebbing5@rakuten.co.jp  Genderfluid    77
7       Wendy     Popland                wpopland6@cbc.ca  Genderfluid    67
8         Nye   Anthiftle        nanthiftle7@examiner.com      Agender    99
9         Der      Neames        dneames8@ycombinator.com     Bigender    77
10      David    Meredyth      dmeredyth9@telegraph.co.uk       Female    68
11     Elijah       Heine                eheinea@imdb.com     Bigender    85
12  Raffaello   Firminger       rfirmingerb@bloomberg.com       Female    74
13    Kristan     Abriani              kabrianic@phoca.cz   Polygender    29
14    Mariana     Shevels               mshevelsd@msn.com      Agender    60
15       Suzy     Bortoli             sbortolie@opera.com      Agender   100

Katta qiymatlarni tashlab yuboramiz:

In [9]:
idx = df[df.age>123].index
df.drop(index=idx, inplace=True)
df.describe()

age
count   93.000000
mean    52.612903
std     29.978265
min      1.000000
25%     29.000000
50%     52.000000
75%     77.000000
max    100.000000

Ko'rib turganingizdek, jadval qiymatlari normal holatga qaytdi.

Jadvlada yana bir e'tiborimni tortgna jihati, jinslar faqatgina erkak (`Male`) va ayol (`Female`) emas.

In [10]:
df.gender.value_counts()

gender
Male           21
Bigender       12
Female         11
Genderfluid    11
Non-binary     10
Genderqueer    10
Agender         9
Polygender      9
Name: count, dtype: int64

Agar bu sizning keyingi hisoblaringizga ta'sir o'tkazadi desangiz, bu qiymatlarni ham filtrlab olishingiz mumkin:

In [11]:
df[df['gender'].isin(['Male','Female'])]

first_name     last_name                         email  gender  age
id                                                                    
2     Pietrek       Scanlin        pscanlin1@yolasite.com    Male   91
5    Mohammed    Habbershon  mhabbershon4@sourceforge.net  Female   48
10      David      Meredyth    dmeredyth9@telegraph.co.uk  Female   68
12  Raffaello     Firminger     rfirmingerb@bloomberg.com  Female   74
17      Byram   Martynikhin     bmartynikhing@jiathis.com    Male   40
19      Luisa     Bradborne          lbradbornei@fema.gov  Female   75
22        Ray         Padly           rpadlyl@twitpic.com    Male   85
26   Ephrayim    Westmacott         ewestmacottp@ning.com    Male   77
29      Biron      Acklands        backlandss@comsenz.com    Male   13
33   Ebenezer       Greeves         egreevesw@myspace.com    Male   26
34     Brooks          Vose             bvosex@amazon.com  Female   34
36        Axe       Arundel           aarundelz@yahoo.com  Female   14
43       Bili     Swetenham        bswetenham16@europa.eu    Male   23
46     Alvira  Stollsteiner     astollsteiner19@mysql.com    Male   66
58     Brynna        Worsom              bworsom1l@wp.com    Male   66
61       Page       Merrell     pmerrell1o@slideshare.net    Male    1
62     Kennan         Visco          kvisco1p@jiathis.com  Female   94
63   Adrianne       Gommowe  agommowe1q@deliciousdays.com    Male  100
67     Bunnie        Ozanne      bozanne1u@whitehouse.gov    Male   92
70      Ambur        Aickin           aaickin1x@about.com    Male   60
72   Eleanora       Kamenar       ekamenar1z@examiner.com    Male   47
74      Curry    Smalecombe  csmalecombe21@washington.edu    Male   69
76      Misty         Blanc         mblanc23@amazon.co.jp    Male   74
77     Rudolf        Kalewe           rkalewe24@fotki.com    Male   12
81       Alex         Darth    adarth28@ezinearticles.com  Female   98
86    Rogerio          Akid           rakid2d@latimes.com  Female    3
87      Peder        Murrie       pmurrie2e@bloglovin.com    Male   69
88       Kain     Dumphries  kdumphries2f@hugedomains.com    Male   93
90  Zachariah         Meins          zmeins2h@weather.com    Male   69
91      Judas        Maffin         jmaffin2i@dropbox.com    Male    3
94     Sheree        Cockin     scockin2l@accuweather.com  Female   94
97     Kattie       Claxson        kclaxson2o@booking.com  Female   18

### Jadvlani aralashtirib tashlash

Ba'zida, tadqiqod uchun jadvaldagi qiymatlar o'rnini aralashtirib tashlash talab qilinishi mumkin. Misol uchun, sizda ma'lumotlar ma'lum bir tartibda turibdi, lekin siz ma'lumotlarni tasodifiy tartibda chiqarmoqchiisiz, buning uchun `numpy` kutubxonasining `random` modulidagi `permutation` funksiyasiga murojaat qilamiz.

Bizda quyidagi jadval bor:

In [12]:
df.head()

first_name   last_name                         email       gender  age
id                                                                       
1     Dorothy     Verryan             dverryan0@who.int   Non-binary    1
2     Pietrek     Scanlin        pscanlin1@yolasite.com         Male   91
4        Knox  Cunningham  kcunningham3@yellowpages.com  Genderfluid    6
5    Mohammed  Habbershon  mhabbershon4@sourceforge.net       Female   48
6       Glynn      Ebbing        gebbing5@rakuten.co.jp  Genderfluid   77

Keling, jadval uzunligini topaylik:

In [13]:
len(df)

93

Jadvalimizda 93 ta qator mavjud, biz 1-93 gacha sonlarni tasodifiy tartibda saqlovchi ro'yxat yaratamiz:

In [20]:
randlist = np.random.permutation(93)
randlist

array([20, 31, 15,  9, 26,  8, 36,  2, 17, 65, 84, 29, 58, 39, 47, 19, 23,
       64, 30, 54, 88, 92, 14, 22, 74, 57, 77, 52,  7, 72, 82, 16, 62, 56,
        5, 66, 40, 89, 90, 11, 78, 34, 76, 35, 68, 21, 80, 86, 91, 70, 87,
       38, 85, 41, 45, 13, 73, 59, 48, 33, 43, 60, 25, 46, 51,  3, 55, 27,
       83,  0, 67, 63, 61, 69, 18,  4, 44, 42, 53, 32, 75, 71, 37, 10,  1,
       24,  6, 12, 49, 50, 79, 81, 28])

Navbat, `df` ni yuqoridagi tasodifiy tartibda tahlash, buning uchun `.take()` meotidga murojaat qilamiz

In [21]:
df.take(randlist)

first_name    last_name                         email       gender  age
id                                                                        
22        Ray        Padly           rpadlyl@twitpic.com         Male   85
33   Ebenezer      Greeves         egreevesw@myspace.com         Male   26
17      Byram  Martynikhin     bmartynikhing@jiathis.com         Male   40
11     Elijah        Heine              eheinea@imdb.com     Bigender   85
28    Philipa         Dron          pdronr@people.com.cn  Genderqueer   52
..        ...          ...                           ...          ...  ...
54    Gregory        Criag       gcriag1h@purevolume.com   Polygender   50
55      Hiram     Bringloe            hbringloe1i@pen.io   Polygender   50
86    Rogerio         Akid           rakid2d@latimes.com       Female    3
88       Kain    Dumphries  kdumphries2f@hugedomains.com         Male   93
30       Linn    Mc Mechan        lmcmechant@sina.com.cn  Genderfluid   33

[93 rows x 5 columns]

Marhamat, jadvalimiz tasodifiy tartibda tahlandi.

### Jadvladan  tasodifiy qiymatlarni chiqarib olish.
Buning uchun bizga avvaldan tanish .sample() metodiga murojaat qilamiz va aynan nechta qiymat kerak ekanini ham ko'rsatamiz.

Bunda tasodifiy qatorlar **takrorlanmagan** holda chiqadi.

In [22]:
df.sample(5)

first_name last_name                         email      gender  age
id                                                                    
51  Gwendolin   Kennion  gkennion1e@independent.co.uk    Bigender   83
13    Kristan   Abriani            kabrianic@phoca.cz  Polygender   29
70      Ambur    Aickin           aaickin1x@about.com        Male   60
18    Artemus  Shovelin         ashovelinh@sphinn.com  Non-binary   58
2     Pietrek   Scanlin        pscanlin1@yolasite.com        Male   91

Agar qatorlar takrorlanishini istasangiz, `replace=True` parametrini ko'rsatasiz:

In [25]:
df.sample(10, replace=True)

first_name  last_name                        email       gender  age
id                                                                      
58      Brynna     Worsom             bworsom1l@wp.com         Male   66
1      Dorothy    Verryan            dverryan0@who.int   Non-binary    1
30        Linn  Mc Mechan       lmcmechant@sina.com.cn  Genderfluid   33
71        Zane   Brickham   zbrickham1y@soundcloud.com  Genderqueer    5
84       Tildi   Phythien       tphythien2b@ustream.tv  Genderfluid   76
57    Salvador    Dusting        sdusting1k@tripod.com   Non-binary   96
2      Pietrek    Scanlin       pscanlin1@yolasite.com         Male   91
48   Sutherlan     Duigan            sduigan1b@ask.com   Non-binary   27
50  Bartolomeo    Heffron      bheffron1d@bluehost.com      Agender    9
52      Tillie   Winslett  twinslett1f@theatlantic.com   Non-binary   90

Ahamiyat bering, `14` indeks ikki marta chiqdi.